core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java

```diff
@@ -1336,6 +1338,132 @@ public TrimResult trimFields(
     return result(newSnapshot, inputMapping, snapshot);
   }
 
+  /**
+   * Trims {@link LogicalCorrelate} nodes.
+   */
+  public TrimResult trimFields(LogicalCorrelate correlate,
+      ImmutableBitSet fieldsUsed, Set<RelDataTypeField> extraFields) {
+    if (!extraFields.isEmpty()) {
+      // bail out with generic trim
+      return trimFields((RelNode) correlate, fieldsUsed, extraFields);
+    }
+
+    fieldsUsed = fieldsUsed.union(correlate.getRequiredColumns());
+
+    List<RelNode> newInputs = new ArrayList<>();
+    List<Mapping> inputMappings = new ArrayList<>();
+    int changeCount = 0;
+    int offset = 0;
+    for (RelNode input : correlate.getInputs()) {
+      final RelDataType inputRowType = input.getRowType();
+      final int inputFieldCount = inputRowType.getFieldCount();
+
+      ImmutableBitSet currentInputFieldsUsed = fieldsUsed
+          .intersect(ImmutableBitSet.range(offset, offset + inputFieldCount))
+          .shift(-offset);
+
+      TrimResult trimResult =
+          dispatchTrimFields(input, currentInputFieldsUsed, extraFields);
+
+      newInputs.add(trimResult.left);
+      inputMappings.add(trimResult.right);
+
+      offset += inputFieldCount;
+
+      if (trimResult.left != input) {
+        changeCount++;
+      }
+    }
+
+    if (changeCount == 0) {
+      return result(correlate,
+          Mappings.createIdentity(correlate.getRowType().getFieldCount()));
+    }
+
+    Mapping mapping = makeMapping(inputMappings);
+    RexBuilder rexBuilder = relBuilder.getRexBuilder();
+
+    RexCorrelVariableMapShuttle rexVisitor =
+        new RexCorrelVariableMapShuttle(correlate.getCorrelationId(),
+            newInputs.get(0).getRowType(), mapping, rexBuilder);
+    final LogicalCorrelate newCorrelate =
+        correlate
+            .copy(correlate.getTraitSet(),
+                newInputs.get(0),
+                newInputs.get(1).accept(new RexRewritingRelShuttle(rexVisitor)),
```

Comment:
I find it easier to debug if I avoid complex expressions - I would use a temporary for the result of the visitor.

Reply:
this also assumes implicitly that newInputs.size() == 2; perhaps this is obvious, but then it would not harm to assert that this holds for the correlate

Reply:
extracted them to be `newLeft` and `newRight` (I was thinking the same earlier ; but I've only extracted the `rexVistior`)

`LogicalCorrelate`'s constructor has only `left` and `right` - but that may change...I've added a check to ensure that this only applies when `correlate.getInputs().size()` is 2